src/GafferScene/EvaluateLightLinks.cpp

```diff
@@ -0,0 +1,127 @@
+//        disclaimer.
+//
+//      * Redistributions in binary form must reproduce the above
+//        copyright notice, this list of conditions and the following
+//        disclaimer in the documentation and/or other materials provided with
+//        the distribution.
+//
+//      * Neither the name of John Haddon nor the names of
+//        any other contributors to this software may be used to endorse or
+//        promote products derived from this software without specific prior
+//        written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include <cstdio>
+
+#include "IECore/CompoundData.h"
+
+#include "GafferScene/EvaluateLightLinks.h"
+#include "GafferScene/SetAlgo.h"
+
+using namespace IECore;
+using namespace Gaffer;
+using namespace GafferScene;
+
+IE_CORE_DEFINERUNTIMETYPED( EvaluateLightLinks );
+
+EvaluateLightLinks::EvaluateLightLinks( const std::string &name )
+	:	SceneProcessor( name )
+{
+	m_lightLinkAttrName = "gaffer:linkedLights";
+
+	// Fast pass-throughs for the things we don't alter.
+	outPlug()->boundPlug()->setInput( inPlug()->boundPlug() );
+	outPlug()->transformPlug()->setInput( inPlug()->transformPlug() );
+	outPlug()->setNamesPlug()->setInput( inPlug()->setNamesPlug() );
+	outPlug()->setPlug()->setInput( inPlug()->setPlug() );
+	outPlug()->objectPlug()->setInput( inPlug()->objectPlug() );
+	outPlug()->childNamesPlug()->setInput( inPlug()->childNamesPlug() );
+	outPlug()->globalsPlug()->setInput( inPlug()->globalsPlug() );
+}
+
+EvaluateLightLinks::~EvaluateLightLinks()
+{
+}
+
+void EvaluateLightLinks::affects( const Gaffer::Plug *input, AffectedPlugsContainer &outputs ) const
+{
+	SceneProcessor::affects( input, outputs );
+
+	if( input == inPlug()->attributesPlug() || input == inPlug()->setPlug() || input == inPlug()->globalsPlug() )
+	{
+		outputs.push_back( outPlug()->attributesPlug() );
+	}
+}
+
+void EvaluateLightLinks::hashAttributes( const ScenePath &path, const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const
+{
+	SceneProcessor::hashAttributes( path, context, parent, h );
+
+	inPlug()->attributesPlug()->hash( h );
```

Comment:
I think we can be friendlier to Gaffer's cache here - by passing through an unchanged hash when the light links attribute doesn't exist we can allow Gaffer to share cache entries with our input for anything we don't touch. Something like this (untested!)...

```
const MurmurHash inputHash = inPlug()->attributesPlug()->hash();
ConstCompoundObjectPtr attributes = inPlug()->attributesPlug()->getValue( & inputHash );
ConstStringDataPtr expressionData = attributes->member<StringData>( "linkedLights" );
if( !expressionData )
{
    // Pass through.
    h = inputHash;
    return;
}

SceneProcessor::hashAttributes( path, context, parent, h );
h.append( inputHash );
h.append( SetAlgo::setExpressionHash( expressionData->readable(), inPlug() );
```